src/main/java/net/neoforged/neoforge/common/data/DataMapProvider.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.data;
+
+import com.google.gson.JsonElement;
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.JsonOps;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import net.minecraft.core.Holder;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.core.Registry;
+import net.minecraft.data.CachedOutput;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.PackOutput;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagKey;
+import net.neoforged.neoforge.common.conditions.ConditionalOps;
+import net.neoforged.neoforge.common.conditions.ICondition;
+import net.neoforged.neoforge.common.conditions.WithConditions;
+import net.neoforged.neoforge.registries.DataMapLoader;
+import net.neoforged.neoforge.registries.datamaps.AdvancedDataMapType;
+import net.neoforged.neoforge.registries.datamaps.DataMapEntry;
+import net.neoforged.neoforge.registries.datamaps.DataMapFile;
+import net.neoforged.neoforge.registries.datamaps.DataMapType;
+import net.neoforged.neoforge.registries.datamaps.DataMapValueRemover;
+
+/**
+ * A provider for {@link DataMapType data map} generation.
+ */
+public abstract class DataMapProvider implements DataProvider {
+    protected final CompletableFuture<HolderLookup.Provider> lookupProvider;
+    protected final PackOutput.PathProvider pathProvider;
+    private final Map<DataMapType<?, ?>, Builder<?, ?>> builders = new HashMap<>();
+
+    /**
+     * Create a new provider.
+     *
+     * @param packOutput     the output location
+     * @param lookupProvider a {@linkplain CompletableFuture} supplying the registries
+     */
+    protected DataMapProvider(PackOutput packOutput, CompletableFuture<HolderLookup.Provider> lookupProvider) {
+        this.lookupProvider = lookupProvider;
+        this.pathProvider = packOutput.createPathProvider(PackOutput.Target.DATA_PACK, DataMapLoader.PATH);
+    }
+
+    @Override
+    public CompletableFuture<?> run(CachedOutput cache) {
+        gather();
+
+        return lookupProvider.thenCompose(provider -> {
+            final DynamicOps<JsonElement> dynamicOps = ConditionalOps.create(RegistryOps.create(JsonOps.INSTANCE, provider), ICondition.IContext.EMPTY);
```

Comment:
Registry ops is enough, the conditional codec should default to an `EMPTY` context.